spec/integration/client_side_encryption/corpus_spec.rb

```diff
@@ -0,0 +1,231 @@
+        if ['_id', 'altname_aws', 'altname_local'].include?(key)
+          corpus_copied[key] = doc
+          next
+        end
+
+        if doc['method'] == 'auto'
+          corpus_copied[key] = doc
+        elsif doc['method'] == 'explicit'
+          options = if doc['identifier'] == 'id'
+            key_id = if doc['kms'] == 'local'
+              'LOCALAAAAAAAAAAAAAAAAA=='
+            else
+              'AWSAAAAAAAAAAAAAAAAAAA=='
+            end
+
+            { key_id: BSON::Binary.new(Base64.decode64(key_id), :uuid) }
+          elsif doc['identifier'] == 'altname'
+            { key_alt_name: doc['kms'] }
+          end
+
+          algorithm = if doc['algo'] == 'rand'
+            'AEAD_AES_256_CBC_HMAC_SHA_512-Random'
+          else
+            'AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic'
+          end
+
+          begin
+            encrypted_value = client_encryption.encrypt(
+              doc['value'],
+              options.merge({ algorithm: algorithm })
+            )
+
+            corpus_copied[key] = doc.merge('value' => encrypted_value)
+          rescue => e
+            # If doc['allowed'] is true, it means that this field should have
+            # been encrypted without error, and thus that this error is unexpected.
+            # If doc['allowed'] is false, this error was expected and the value
+            # should be copied over without being encrypted.
+            if doc['allowed']
+              raise "Unexpected error occured in client-side encryption " +
+                "corpus tests: #{e.class}, #{e.message}"
+            end
+
+            corpus_copied[key] = doc
+          end
+        end
+      end
+
+      corpus_copied
+    end
+
+    before do
+      client.use(:db)[:coll].drop
+
+      client.use(:admin)[:datakeys].drop
+      client.use(:admin)[:datakeys].insert_one(local_data_key)
+      client.use(:admin)[:datakeys].insert_one(aws_data_key)
+    end
+
+    shared_context 'with jsonSchema collection validator' do
+      let(:local_schema_map) { nil }
+
+      before do
+        client.use(:db)[:coll,
+          {
+            'validator' => { '$jsonSchema' => test_schema_map }
+          }
+        ].create
+      end
+    end
+
+    shared_context 'with local schema map' do
+      let(:local_schema_map) { { 'db.coll' => test_schema_map } }
+    end
+
+    shared_examples 'a functioning encrypter' do
+      it 'properly encrypts and decrypts a document' do
+        corpus_encrypted_id = client_encrypted[:coll]
+          .insert_one(corpus_copied)
+          .inserted_id
+
+        corpus_decrypted = client_encrypted[:coll]
+          .find(_id: corpus_encrypted_id)
+          .first
+
+        # Ensure that corpus_decrypted is the same as the original corpus
+        # document by checking that they have the same set of keys, and that
+        # they have the same values at those keys (improved diagnostics).
+        expect(corpus_decrypted.keys - corpus.keys).to be_empty
```

Comment:
This allows corpus_decrypted to omit keys in corpus. If the set of keys allowed to be omitted is known, can those be explicitly removed from `corpus` and then an exact comparison performed between the adjusted `corpus` and `corpus_decrypted`?